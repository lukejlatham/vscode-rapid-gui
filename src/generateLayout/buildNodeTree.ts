import { z } from "zod";
import {
  backgroundNodeLayout,
  nodeTreeRootSchema,
  craftjsNodeSchema,
  buttonSchema,
  sectionsSchema,
  checkboxesSchema,
  containerSchema,
  inputSchema,
  labelSchema,
  radioButtonSchema,
  textBoxSchema,
  iconSchema,
  imageSchema,
  gridCellSchema,
  textSchema,
  ThemedLayoutSchema,
  dropdownSchema,
  sliderSchema,
  fontList,
  themeList,
} from "../../webview-ui/src/types";
import { adjustLayoutToFitGrid } from "./gridLayoutCorrection";
import { applyThemeToSchema } from "./applyTheming";

type BackgroundType = z.infer<typeof backgroundNodeLayout>;
type NodeTreeRootType = z.infer<typeof nodeTreeRootSchema>;
type SectionsSchema = z.infer<typeof sectionsSchema>;
type NodeSection = z.infer<typeof craftjsNodeSchema>;

const elementSchemas = {
  Button: buttonSchema,
  Checkboxes: checkboxesSchema,
  Input: inputSchema,
  Label: labelSchema,
  RadioButtons: radioButtonSchema,
  TextBox: textBoxSchema,
  Icon: iconSchema,
  Image: imageSchema,
  Text: textSchema,
  Dropdown: dropdownSchema,
  Slider: sliderSchema,
};

interface LayoutDimensions {
  rows: number;
  columns: number;
  ids: string[];
}

function calculateLayoutDimensions(layout: SectionsSchema): LayoutDimensions {
  let maxX = 0;
  let maxY = 0;
  const ids: string[] = layout.map((section) => {
    const sectionRight = section.xPosition + section.width;
    const sectionBottom = section.yPosition + section.height;

    if (sectionRight > maxX) maxX = sectionRight;
    if (sectionBottom > maxY) maxY = sectionBottom;

    return section.section;
  });

  return { rows: 10, columns: 10, ids };
}

function createNode(
  resolvedName: string,
  isCanvas: boolean,
  parent: string,
  custom: { [key: string]: any } = {},
  props: { [key: string]: any } = {},
  children: string[] = []
): NodeSection {
  return {
    type: { resolvedName },
    isCanvas,
    props,
    displayName: resolvedName,
    custom,
    parent,
    hidden: false,
    nodes: children,
    linkedNodes: {},
  };
}

function calculateImageScale(containerWidth: number, containerHeight: number): number {
  const aspectRatio = containerWidth / containerHeight;

  if (aspectRatio >= 1) {
    // Landscape or square container
    return containerHeight / containerWidth;
  } else {
    // Portrait container
    return containerWidth / containerHeight;
  }
}

function generateSectionNodes(sections: ThemedLayoutSchema[]): { [key: string]: NodeSection } {
  const nodes: { [key: string]: NodeSection } = {};

  sections.forEach((section, index) => {
    const sanitizedSectionName = section.section.replace(/\s+/g, "");
    const gridCellId = `${sanitizedSectionName}GridCell`;
    const containerId = `${sanitizedSectionName}Container`;

    nodes[gridCellId] = createNode(
      "GridCell",
      true,
      "ROOT",
      { id: index.toString() },
      gridCellSchema.parse({}),
      [containerId]
    );

    const flexDirectionOverride = section.width > section.height ? "row" : "column";

    nodes[containerId] = createNode(
      "Container",
      true,
      gridCellId,
      {},
      containerSchema.parse({
        flexDirection: flexDirectionOverride,
        backgroundColor: section.backgroundColor,
        borderColor: section.borderColor,
        width: 100,
        height: 100,
        shadowColor: "transparent",
        shadowOffsetX: 1,
        shadowOffsetY: 1,
        shadowBlur: 1,
      }),
      []
    );

    const childIds = section.children.map((child, childIndex) => {
      const childId = `${sanitizedSectionName}${child.element}${childIndex}`;
      const schema = elementSchemas[child.element as keyof typeof elementSchemas];

      if (!schema) {
        throw new Error(`Unknown element type: ${child.element}`);
      }

      let childProps = schema.parse(child);

      if (child.element === "Image") {
        const scaledWidth = Math.floor(calculateImageScale(section.width, section.height) * 40);
        childProps = { ...childProps, width: scaledWidth };
      }

      if (child.element === "Checkboxes" || child.element === "RadioButtons") {
        childProps = { ...childProps, direction: flexDirectionOverride };
      }

      nodes[childId] = createNode(child.element, false, containerId, {}, childProps);
      return childId;
    });

    nodes[containerId].nodes = childIds;
  });

  return nodes;
}

function createBackgroundNode(
  dimensions: LayoutDimensions,
  layout: BackgroundType[],
  backgroundColor: string
): NodeTreeRootType {
  const linkedNodes = dimensions.ids.reduce((acc, id, index) => {
    const sanitizedId = id.replace(/\s+/g, "");
    acc[String(index)] = sanitizedId + "GridCell";
    return acc;
  }, {} as Record<string, string>);

  return {
    type: { resolvedName: "Background" },
    isCanvas: true,
    props: {
      rows: dimensions.rows,
      columns: dimensions.columns,
      lockedGrid: true,
      backgroundColor,
      layout,
    },
    displayName: "Background",
    custom: {},
    parent: null,
    hidden: false,
    nodes: [],
    linkedNodes,
  };
}

function buildNodeTree(
  generatedLayout: SectionsSchema,
  themeName: string,
  chosenFont: string
): string {
  const initialGrid = generatedLayout.map((section, index) => ({
    i: String(index),
    x: section.xPosition,
    y: section.yPosition,
    w: section.width,
    h: section.height,
  }));

  const adjustedLayout = adjustLayoutToFitGrid(initialGrid);

  const updatedGeneratedLayout = generatedLayout.map((section, index) => {
    const adjustedCell = adjustedLayout.find((cell) => cell.i === String(index));
    if (adjustedCell) {
      return {
        ...section,
        xPosition: adjustedCell.x,
        yPosition: adjustedCell.y,
        width: adjustedCell.w,
        height: adjustedCell.h,
      };
    }
    return section;
  });

  const layoutDimensions = calculateLayoutDimensions(updatedGeneratedLayout);

  const chosenTheme = themeList.find(
    (theme) => theme.themeGenerationName.toLowerCase() === themeName.toLowerCase()
  );

  if (!chosenTheme) {
    throw new Error(`Theme "${chosenTheme}" not found`);
  }

  const themedNodes = applyThemeToSchema(updatedGeneratedLayout, chosenTheme.scheme);

  const sectionNodes = generateSectionNodes(themedNodes);

  const font = fontList.find((font) => font.name === chosenFont);

  const selectedFont = font.displayName;

  if (selectedFont) {
    Object.values(sectionNodes).forEach((node) => {
      if ("fontFamily" in node.props) {
        node.props.fontFamily = `${selectedFont}`;
      }
    });
  }

  console.log("applying shadows?", chosenTheme.scheme.shadows);

  const shadows = chosenTheme.scheme.shadows;

  if (shadows) {
    Object.values(sectionNodes).forEach((node) => {
      if ("shadowColor" in node.props) {
        node.props.shadowColor = "rgba(0, 0, 0, 0.6)";
        node.props.shadowOffsetX = 2;
        node.props.shadowOffsetY = 2;
        node.props.shadowBlur = 2;
      }
    });
  }

  const backgroundNode = createBackgroundNode(
    layoutDimensions,
    adjustedLayout,
    chosenTheme.scheme.backgroundColor
  );

  const combinedNodes = { ROOT: backgroundNode, ...sectionNodes };

  const stringifiedNodes = JSON.stringify(combinedNodes);

  return stringifiedNodes;
}

export { buildNodeTree, calculateLayoutDimensions, generateSectionNodes, createBackgroundNode };
