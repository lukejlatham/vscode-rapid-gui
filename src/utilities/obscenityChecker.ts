import { RegExpMatcher, englishDataset, englishRecommendedTransformers } from "obscenity";

const matcher = new RegExpMatcher({
  ...englishDataset.build(),
  ...englishRecommendedTransformers,
});

export function checkForObscenities(content: string): void {
  if (matcher.hasMatch(content)) {
    throw new Error("Potentially offensive content detected. Operation aborted.");
  }
}

export function warnAboutImageContent(): void {
  console.warn(
    "Please ensure the uploaded image does not contain any inappropriate or offensive content."
  );
}
