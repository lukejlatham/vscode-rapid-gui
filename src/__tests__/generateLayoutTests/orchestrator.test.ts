import { OpenAI } from "openai";
import { generateFromText } from "../../generateLayout/generateLayout";
import { buildNodeTree } from "../../generateLayout/buildNodeTree";
import { checkForObscenities } from "../../utilities/obscenityChecker";
import { layoutSchema } from "../../../webview-ui/src/types";

jest.mock("openai");
jest.mock("../../utilities/obscenityChecker");

describe("Layout Generation Orchestrator Integration Test", () => {
  let mockOpenAIClient: jest.Mocked<OpenAI>;

  beforeEach(() => {
    // Mock OpenAI client
    mockOpenAIClient = {
      beta: { chat: { completions: { parse: jest.fn() } as any } },
    } as any;

    // Mock obscenity checker
    (checkForObscenities as jest.Mock).mockImplementation(() => true);
  });

  it("should generate a layout, transform it into a node tree, and validate it", async () => {
    // Mock layout returned by OpenAI API (with fontSize as string before transformation)
    const mockGeneratedLayout = {
      theme: "microsoftWordBlue",
      fontFamily: "IBM",
      sections: [
        {
          section: "Header",
          xPosition: 0,
          yPosition: 0,
          width: 10,
          height: 1,
          flexDirection: "row",
          backgroundColor: "DarkAccent",
          children: [
            {
              element: "Label",
              text: "Microsoft Word",
              bold: true,
              fontColor: "Main",
              fontSize: "Title",
            },
          ],
        },
        {
          section: "Sidebar",
          xPosition: 0,
          yPosition: 1,
          width: 1,
          height: 9,
          flexDirection: "column",
          backgroundColor: "Main",
          children: [
            { element: "Icon", vscIcon: "document", iconSize: 32, iconColor: "DarkAccent" },
            { element: "Icon", vscIcon: "folder", iconSize: 32, iconColor: "DarkAccent" },
            { element: "Icon", vscIcon: "settings", iconSize: 32, iconColor: "DarkAccent" },
          ],
        },
        {
          section: "Main Document Area",
          xPosition: 1,
          yPosition: 1,
          width: 7,
          height: 7,
          flexDirection: "column",
          backgroundColor: "LightAccent",
          children: [
            { element: "Text", fontColor: "DarkAccent" },
            { element: "Input", placeholder: "Type your document here..." },
          ],
        },
        {
          section: "Page Layout",
          xPosition: 8,
          yPosition: 1,
          width: 2,
          height: 3,
          flexDirection: "column",
          backgroundColor: "Main",
          children: [{ element: "RadioButtons", optionLabels: ["Portrait", "Landscape"] }],
        },
        {
          section: "Font Settings",
          xPosition: 8,
          yPosition: 4,
          width: 2,
          height: 2,
          flexDirection: "column",
          backgroundColor: "Main",
          children: [
            { element: "Dropdown", optionLabels: ["Arial", "Times New Roman", "Calibri"] },
            { element: "Slider", header: "Font Size", backgroundColor: "LightAccent" },
          ],
        },
        {
          section: "Footer",
          xPosition: 1,
          yPosition: 8,
          width: 9,
          height: 2,
          flexDirection: "row",
          backgroundColor: "DarkAccent",
          children: [
            {
              element: "Label",
              text: "Status: Ready",
              bold: false,
              fontColor: "Main",
              fontSize: "Label",
            },
          ],
        },
      ],
    };

    (mockOpenAIClient.beta.chat.completions.parse as jest.Mock).mockResolvedValue({
      choices: [{ message: { parsed: mockGeneratedLayout } }],
      usage: { prompt_tokens: 100, completion_tokens: 200 },
    });

    // Step 1: Generate the layout using the mocked OpenAI API
    const layout = await generateFromText(mockOpenAIClient, "Create a layout for VS Code");

    // Ensure layout is parsed via Zod schema before passing to buildNodeTree
    const parsedLayout = layoutSchema.parse(layout); // This will apply transformations

    // Step 2: Pass the parsed layout through buildNodeTree
    const nodeTree = buildNodeTree(
      parsedLayout.sections,
      parsedLayout.theme,
      parsedLayout.fontFamily
    );

    expect(nodeTree).toMatchSnapshot(); // Snapshot to validate the structure
  });
});
