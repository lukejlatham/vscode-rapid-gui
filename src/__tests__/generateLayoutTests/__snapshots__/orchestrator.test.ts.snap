// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Layout Generation Orchestrator Integration Test should generate a layout, transform it into a node tree, and validate it 1`] = `"{"ROOT":{"type":{"resolvedName":"Background"},"isCanvas":true,"props":{"rows":10,"columns":10,"lockedGrid":true,"backgroundColor":"#FFFFFF","layout":[{"i":"0","x":0,"y":0,"w":10,"h":1},{"i":"1","x":0,"y":1,"w":1,"h":9},{"i":"2","x":1,"y":1,"w":7,"h":7},{"i":"3","x":8,"y":1,"w":2,"h":3},{"i":"4","x":8,"y":4,"w":2,"h":2},{"i":"5","x":1,"y":8,"w":9,"h":2}]},"displayName":"Background","custom":{},"parent":null,"hidden":false,"nodes":[],"linkedNodes":{"0":"HeaderGridCell","1":"SidebarGridCell","2":"MainDocumentAreaGridCell","3":"PageLayoutGridCell","4":"FontSettingsGridCell","5":"FooterGridCell"}},"HeaderGridCell":{"type":{"resolvedName":"GridCell"},"isCanvas":true,"props":{"flexDirection":"row","justifyContent":"center","alignItems":"center","gap":10},"displayName":"GridCell","custom":{"id":"0"},"parent":"ROOT","hidden":false,"nodes":["HeaderContainer"],"linkedNodes":{}},"HeaderContainer":{"type":{"resolvedName":"Container"},"isCanvas":true,"props":{"height":100,"width":100,"flexDirection":"row","justifyContent":"space-around","alignItems":"center","gap":10,"backgroundColor":"#E1DFDD","borderRadius":5,"borderColor":"#C8C6C4","padding":5,"shadowColor":"rgba(0, 0, 0, 0.6)","shadowOffsetX":2,"shadowOffsetY":2,"shadowBlur":2},"displayName":"Container","custom":{},"parent":"HeaderGridCell","hidden":false,"nodes":["HeaderLabel0"],"linkedNodes":{}},"HeaderLabel0":{"type":{"resolvedName":"Label"},"isCanvas":false,"props":{"text":"Microsoft Word","fontSize":32,"fontFamily":"IBM Plex Sans","fontColor":"#252423","userEditable":true,"textAlign":"left","icon":"none","hyperlink":"","bold":true,"italic":false,"underline":false},"displayName":"Label","custom":{},"parent":"HeaderContainer","hidden":false,"nodes":[],"linkedNodes":{}},"SidebarGridCell":{"type":{"resolvedName":"GridCell"},"isCanvas":true,"props":{"flexDirection":"row","justifyContent":"center","alignItems":"center","gap":10},"displayName":"GridCell","custom":{"id":"1"},"parent":"ROOT","hidden":false,"nodes":["SidebarContainer"],"linkedNodes":{}},"SidebarContainer":{"type":{"resolvedName":"Container"},"isCanvas":true,"props":{"height":100,"width":100,"flexDirection":"column","justifyContent":"space-around","alignItems":"center","gap":10,"backgroundColor":"#F3F2F1","borderRadius":5,"borderColor":"#E1DFDD","padding":5,"shadowColor":"rgba(0, 0, 0, 0.6)","shadowOffsetX":2,"shadowOffsetY":2,"shadowBlur":2},"displayName":"Container","custom":{},"parent":"SidebarGridCell","hidden":false,"nodes":["SidebarIcon0","SidebarIcon1","SidebarIcon2"],"linkedNodes":{}},"SidebarIcon0":{"type":{"resolvedName":"Icon"},"isCanvas":false,"props":{"vscIcon":"VscComment","iconSize":32,"iconColor":"#2B579A","hyperlink":""},"displayName":"Icon","custom":{},"parent":"SidebarContainer","hidden":false,"nodes":[],"linkedNodes":{}},"SidebarIcon1":{"type":{"resolvedName":"Icon"},"isCanvas":false,"props":{"vscIcon":"VscFolder","iconSize":32,"iconColor":"#2B579A","hyperlink":""},"displayName":"Icon","custom":{},"parent":"SidebarContainer","hidden":false,"nodes":[],"linkedNodes":{}},"SidebarIcon2":{"type":{"resolvedName":"Icon"},"isCanvas":false,"props":{"vscIcon":"VscSettings","iconSize":32,"iconColor":"#2B579A","hyperlink":""},"displayName":"Icon","custom":{},"parent":"SidebarContainer","hidden":false,"nodes":[],"linkedNodes":{}},"MainDocumentAreaGridCell":{"type":{"resolvedName":"GridCell"},"isCanvas":true,"props":{"flexDirection":"row","justifyContent":"center","alignItems":"center","gap":10},"displayName":"GridCell","custom":{"id":"2"},"parent":"ROOT","hidden":false,"nodes":["MainDocumentAreaContainer"],"linkedNodes":{}},"MainDocumentAreaContainer":{"type":{"resolvedName":"Container"},"isCanvas":true,"props":{"height":100,"width":100,"flexDirection":"column","justifyContent":"space-around","alignItems":"center","gap":10,"backgroundColor":"#FFFFFF","borderRadius":5,"borderColor":"#D2D0CE","padding":5,"shadowColor":"rgba(0, 0, 0, 0.6)","shadowOffsetX":2,"shadowOffsetY":2,"shadowBlur":2},"displayName":"Container","custom":{},"parent":"MainDocumentAreaGridCell","hidden":false,"nodes":["MainDocumentAreaText0","MainDocumentAreaInput1"],"linkedNodes":{}},"MainDocumentAreaText0":{"type":{"resolvedName":"Text"},"isCanvas":false,"props":{"text":"Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat.","fontSize":14,"fontFamily":"IBM Plex Sans","fontColor":"#2B579A","textAlign":"left","bold":false,"italic":false,"underline":false,"hyperlink":"","userEditable":true},"displayName":"Text","custom":{},"parent":"MainDocumentAreaContainer","hidden":false,"nodes":[],"linkedNodes":{}},"MainDocumentAreaInput1":{"type":{"resolvedName":"Input"},"isCanvas":false,"props":{"fontSize":14,"fontFamily":"IBM Plex Sans","fontColor":"#252423","backgroundColor":"var(--colorBrandForeground2Pressed)","borderColor":"var(--colorBrandBackground2)","placeholder":"Type your document here...","borderRadius":4},"displayName":"Input","custom":{},"parent":"MainDocumentAreaContainer","hidden":false,"nodes":[],"linkedNodes":{}},"PageLayoutGridCell":{"type":{"resolvedName":"GridCell"},"isCanvas":true,"props":{"flexDirection":"row","justifyContent":"center","alignItems":"center","gap":10},"displayName":"GridCell","custom":{"id":"3"},"parent":"ROOT","hidden":false,"nodes":["PageLayoutContainer"],"linkedNodes":{}},"PageLayoutContainer":{"type":{"resolvedName":"Container"},"isCanvas":true,"props":{"height":100,"width":100,"flexDirection":"column","justifyContent":"space-around","alignItems":"center","gap":10,"backgroundColor":"#F3F2F1","borderRadius":5,"borderColor":"#E1DFDD","padding":5,"shadowColor":"rgba(0, 0, 0, 0.6)","shadowOffsetX":2,"shadowOffsetY":2,"shadowBlur":2},"displayName":"Container","custom":{},"parent":"PageLayoutGridCell","hidden":false,"nodes":["PageLayoutRadioButtons0"],"linkedNodes":{}},"PageLayoutRadioButtons0":{"type":{"resolvedName":"RadioButtons"},"isCanvas":false,"props":{"header":"","fontFamily":"IBM Plex Sans","numberOfButtons":2,"optionLabels":["Portrait","Landscape"],"fontSize":14,"fontColor":"#252423","direction":"column"},"displayName":"RadioButtons","custom":{},"parent":"PageLayoutContainer","hidden":false,"nodes":[],"linkedNodes":{}},"FontSettingsGridCell":{"type":{"resolvedName":"GridCell"},"isCanvas":true,"props":{"flexDirection":"row","justifyContent":"center","alignItems":"center","gap":10},"displayName":"GridCell","custom":{"id":"4"},"parent":"ROOT","hidden":false,"nodes":["FontSettingsContainer"],"linkedNodes":{}},"FontSettingsContainer":{"type":{"resolvedName":"Container"},"isCanvas":true,"props":{"height":100,"width":100,"flexDirection":"column","justifyContent":"space-around","alignItems":"center","gap":10,"backgroundColor":"#F3F2F1","borderRadius":5,"borderColor":"#E1DFDD","padding":5,"shadowColor":"rgba(0, 0, 0, 0.6)","shadowOffsetX":2,"shadowOffsetY":2,"shadowBlur":2},"displayName":"Container","custom":{},"parent":"FontSettingsGridCell","hidden":false,"nodes":["FontSettingsDropdown0","FontSettingsSlider1"],"linkedNodes":{}},"FontSettingsDropdown0":{"type":{"resolvedName":"Dropdown"},"isCanvas":false,"props":{"header":"","fontFamily":"IBM Plex Sans","optionLabels":["Arial","Times New Roman","Calibri"],"numberOfOptions":2,"fontSize":14,"fontColor":"#252423"},"displayName":"Dropdown","custom":{},"parent":"FontSettingsContainer","hidden":false,"nodes":[],"linkedNodes":{}},"FontSettingsSlider1":{"type":{"resolvedName":"Slider"},"isCanvas":false,"props":{"header":"Font Size","fontFamily":"IBM Plex Sans","min":0,"max":100,"step":1,"fontSize":14,"fontColor":"#252423","backgroundColor":"#41A5EE"},"displayName":"Slider","custom":{},"parent":"FontSettingsContainer","hidden":false,"nodes":[],"linkedNodes":{}},"FooterGridCell":{"type":{"resolvedName":"GridCell"},"isCanvas":true,"props":{"flexDirection":"row","justifyContent":"center","alignItems":"center","gap":10},"displayName":"GridCell","custom":{"id":"5"},"parent":"ROOT","hidden":false,"nodes":["FooterContainer"],"linkedNodes":{}},"FooterContainer":{"type":{"resolvedName":"Container"},"isCanvas":true,"props":{"height":100,"width":100,"flexDirection":"row","justifyContent":"space-around","alignItems":"center","gap":10,"backgroundColor":"#E1DFDD","borderRadius":5,"borderColor":"#C8C6C4","padding":5,"shadowColor":"rgba(0, 0, 0, 0.6)","shadowOffsetX":2,"shadowOffsetY":2,"shadowBlur":2},"displayName":"Container","custom":{},"parent":"FooterGridCell","hidden":false,"nodes":["FooterLabel0"],"linkedNodes":{}},"FooterLabel0":{"type":{"resolvedName":"Label"},"isCanvas":false,"props":{"text":"Status: Ready","fontSize":14,"fontFamily":"IBM Plex Sans","fontColor":"#252423","userEditable":true,"textAlign":"left","icon":"none","hyperlink":"","bold":false,"italic":false,"underline":false},"displayName":"Label","custom":{},"parent":"FooterContainer","hidden":false,"nodes":[],"linkedNodes":{}}}"`;

exports[`Layout Generation Orchestrator Test should generate a layout, transform it into a node tree, and validate it 1`] = `"{"ROOT":{"type":{"resolvedName":"Background"},"isCanvas":true,"props":{"rows":10,"columns":10,"lockedGrid":true,"backgroundColor":"#FFFFFF","layout":[{"i":"0","x":0,"y":0,"w":10,"h":1},{"i":"1","x":0,"y":1,"w":1,"h":9},{"i":"2","x":1,"y":1,"w":7,"h":7},{"i":"3","x":8,"y":1,"w":2,"h":3},{"i":"4","x":8,"y":4,"w":2,"h":2},{"i":"5","x":1,"y":8,"w":9,"h":2}]},"displayName":"Background","custom":{},"parent":null,"hidden":false,"nodes":[],"linkedNodes":{"0":"HeaderGridCell","1":"SidebarGridCell","2":"MainDocumentAreaGridCell","3":"PageLayoutGridCell","4":"FontSettingsGridCell","5":"FooterGridCell"}},"HeaderGridCell":{"type":{"resolvedName":"GridCell"},"isCanvas":true,"props":{"flexDirection":"row","justifyContent":"center","alignItems":"center","gap":10},"displayName":"GridCell","custom":{"id":"0"},"parent":"ROOT","hidden":false,"nodes":["HeaderContainer"],"linkedNodes":{}},"HeaderContainer":{"type":{"resolvedName":"Container"},"isCanvas":true,"props":{"height":100,"width":100,"flexDirection":"row","justifyContent":"space-around","alignItems":"center","gap":10,"backgroundColor":"#E1DFDD","borderRadius":5,"borderColor":"#C8C6C4","padding":5,"shadowColor":"rgba(0, 0, 0, 0.6)","shadowOffsetX":2,"shadowOffsetY":2,"shadowBlur":2},"displayName":"Container","custom":{},"parent":"HeaderGridCell","hidden":false,"nodes":["HeaderLabel0"],"linkedNodes":{}},"HeaderLabel0":{"type":{"resolvedName":"Label"},"isCanvas":false,"props":{"text":"Microsoft Word","fontSize":32,"fontFamily":"IBM Plex Sans","fontColor":"#252423","userEditable":true,"textAlign":"left","icon":"none","hyperlink":"","bold":true,"italic":false,"underline":false},"displayName":"Label","custom":{},"parent":"HeaderContainer","hidden":false,"nodes":[],"linkedNodes":{}},"SidebarGridCell":{"type":{"resolvedName":"GridCell"},"isCanvas":true,"props":{"flexDirection":"row","justifyContent":"center","alignItems":"center","gap":10},"displayName":"GridCell","custom":{"id":"1"},"parent":"ROOT","hidden":false,"nodes":["SidebarContainer"],"linkedNodes":{}},"SidebarContainer":{"type":{"resolvedName":"Container"},"isCanvas":true,"props":{"height":100,"width":100,"flexDirection":"column","justifyContent":"space-around","alignItems":"center","gap":10,"backgroundColor":"#F3F2F1","borderRadius":5,"borderColor":"#E1DFDD","padding":5,"shadowColor":"rgba(0, 0, 0, 0.6)","shadowOffsetX":2,"shadowOffsetY":2,"shadowBlur":2},"displayName":"Container","custom":{},"parent":"SidebarGridCell","hidden":false,"nodes":["SidebarIcon0","SidebarIcon1","SidebarIcon2"],"linkedNodes":{}},"SidebarIcon0":{"type":{"resolvedName":"Icon"},"isCanvas":false,"props":{"vscIcon":"VscComment","iconSize":32,"iconColor":"#2B579A","hyperlink":""},"displayName":"Icon","custom":{},"parent":"SidebarContainer","hidden":false,"nodes":[],"linkedNodes":{}},"SidebarIcon1":{"type":{"resolvedName":"Icon"},"isCanvas":false,"props":{"vscIcon":"VscFolder","iconSize":32,"iconColor":"#2B579A","hyperlink":""},"displayName":"Icon","custom":{},"parent":"SidebarContainer","hidden":false,"nodes":[],"linkedNodes":{}},"SidebarIcon2":{"type":{"resolvedName":"Icon"},"isCanvas":false,"props":{"vscIcon":"VscSettings","iconSize":32,"iconColor":"#2B579A","hyperlink":""},"displayName":"Icon","custom":{},"parent":"SidebarContainer","hidden":false,"nodes":[],"linkedNodes":{}},"MainDocumentAreaGridCell":{"type":{"resolvedName":"GridCell"},"isCanvas":true,"props":{"flexDirection":"row","justifyContent":"center","alignItems":"center","gap":10},"displayName":"GridCell","custom":{"id":"2"},"parent":"ROOT","hidden":false,"nodes":["MainDocumentAreaContainer"],"linkedNodes":{}},"MainDocumentAreaContainer":{"type":{"resolvedName":"Container"},"isCanvas":true,"props":{"height":100,"width":100,"flexDirection":"column","justifyContent":"space-around","alignItems":"center","gap":10,"backgroundColor":"#FFFFFF","borderRadius":5,"borderColor":"#D2D0CE","padding":5,"shadowColor":"rgba(0, 0, 0, 0.6)","shadowOffsetX":2,"shadowOffsetY":2,"shadowBlur":2},"displayName":"Container","custom":{},"parent":"MainDocumentAreaGridCell","hidden":false,"nodes":["MainDocumentAreaText0","MainDocumentAreaInput1"],"linkedNodes":{}},"MainDocumentAreaText0":{"type":{"resolvedName":"Text"},"isCanvas":false,"props":{"text":"Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat.","fontSize":14,"fontFamily":"IBM Plex Sans","fontColor":"#2B579A","textAlign":"left","bold":false,"italic":false,"underline":false,"hyperlink":"","userEditable":true},"displayName":"Text","custom":{},"parent":"MainDocumentAreaContainer","hidden":false,"nodes":[],"linkedNodes":{}},"MainDocumentAreaInput1":{"type":{"resolvedName":"Input"},"isCanvas":false,"props":{"fontSize":14,"fontFamily":"IBM Plex Sans","fontColor":"#252423","backgroundColor":"var(--colorBrandForeground2Pressed)","borderColor":"var(--colorBrandBackground2)","placeholder":"Type your document here...","borderRadius":4},"displayName":"Input","custom":{},"parent":"MainDocumentAreaContainer","hidden":false,"nodes":[],"linkedNodes":{}},"PageLayoutGridCell":{"type":{"resolvedName":"GridCell"},"isCanvas":true,"props":{"flexDirection":"row","justifyContent":"center","alignItems":"center","gap":10},"displayName":"GridCell","custom":{"id":"3"},"parent":"ROOT","hidden":false,"nodes":["PageLayoutContainer"],"linkedNodes":{}},"PageLayoutContainer":{"type":{"resolvedName":"Container"},"isCanvas":true,"props":{"height":100,"width":100,"flexDirection":"column","justifyContent":"space-around","alignItems":"center","gap":10,"backgroundColor":"#F3F2F1","borderRadius":5,"borderColor":"#E1DFDD","padding":5,"shadowColor":"rgba(0, 0, 0, 0.6)","shadowOffsetX":2,"shadowOffsetY":2,"shadowBlur":2},"displayName":"Container","custom":{},"parent":"PageLayoutGridCell","hidden":false,"nodes":["PageLayoutRadioButtons0"],"linkedNodes":{}},"PageLayoutRadioButtons0":{"type":{"resolvedName":"RadioButtons"},"isCanvas":false,"props":{"header":"","fontFamily":"IBM Plex Sans","numberOfButtons":2,"optionLabels":["Portrait","Landscape"],"fontSize":14,"fontColor":"#252423","direction":"column"},"displayName":"RadioButtons","custom":{},"parent":"PageLayoutContainer","hidden":false,"nodes":[],"linkedNodes":{}},"FontSettingsGridCell":{"type":{"resolvedName":"GridCell"},"isCanvas":true,"props":{"flexDirection":"row","justifyContent":"center","alignItems":"center","gap":10},"displayName":"GridCell","custom":{"id":"4"},"parent":"ROOT","hidden":false,"nodes":["FontSettingsContainer"],"linkedNodes":{}},"FontSettingsContainer":{"type":{"resolvedName":"Container"},"isCanvas":true,"props":{"height":100,"width":100,"flexDirection":"column","justifyContent":"space-around","alignItems":"center","gap":10,"backgroundColor":"#F3F2F1","borderRadius":5,"borderColor":"#E1DFDD","padding":5,"shadowColor":"rgba(0, 0, 0, 0.6)","shadowOffsetX":2,"shadowOffsetY":2,"shadowBlur":2},"displayName":"Container","custom":{},"parent":"FontSettingsGridCell","hidden":false,"nodes":["FontSettingsDropdown0","FontSettingsSlider1"],"linkedNodes":{}},"FontSettingsDropdown0":{"type":{"resolvedName":"Dropdown"},"isCanvas":false,"props":{"header":"","fontFamily":"IBM Plex Sans","optionLabels":["Arial","Times New Roman","Calibri"],"numberOfOptions":2,"fontSize":14,"fontColor":"#252423"},"displayName":"Dropdown","custom":{},"parent":"FontSettingsContainer","hidden":false,"nodes":[],"linkedNodes":{}},"FontSettingsSlider1":{"type":{"resolvedName":"Slider"},"isCanvas":false,"props":{"header":"Font Size","fontFamily":"IBM Plex Sans","min":0,"max":100,"step":1,"fontSize":14,"fontColor":"#252423","backgroundColor":"#41A5EE"},"displayName":"Slider","custom":{},"parent":"FontSettingsContainer","hidden":false,"nodes":[],"linkedNodes":{}},"FooterGridCell":{"type":{"resolvedName":"GridCell"},"isCanvas":true,"props":{"flexDirection":"row","justifyContent":"center","alignItems":"center","gap":10},"displayName":"GridCell","custom":{"id":"5"},"parent":"ROOT","hidden":false,"nodes":["FooterContainer"],"linkedNodes":{}},"FooterContainer":{"type":{"resolvedName":"Container"},"isCanvas":true,"props":{"height":100,"width":100,"flexDirection":"row","justifyContent":"space-around","alignItems":"center","gap":10,"backgroundColor":"#E1DFDD","borderRadius":5,"borderColor":"#C8C6C4","padding":5,"shadowColor":"rgba(0, 0, 0, 0.6)","shadowOffsetX":2,"shadowOffsetY":2,"shadowBlur":2},"displayName":"Container","custom":{},"parent":"FooterGridCell","hidden":false,"nodes":["FooterLabel0"],"linkedNodes":{}},"FooterLabel0":{"type":{"resolvedName":"Label"},"isCanvas":false,"props":{"text":"Status: Ready","fontSize":14,"fontFamily":"IBM Plex Sans","fontColor":"#252423","userEditable":true,"textAlign":"left","icon":"none","hyperlink":"","bold":false,"italic":false,"underline":false},"displayName":"Label","custom":{},"parent":"FooterContainer","hidden":false,"nodes":[],"linkedNodes":{}}}"`;
